### DISCUSSION TITLE
"Datasource <default> enables XA but transaction recovery is  not enabled"

### DISCUSSION THREAD

**Discussion by febeling** [keycloak/keycloak#23385] [Q&A]
I'm running 22.0.3 against PostgresQL 15. During startup this warning message is written to the log:
```
WARN  [io.quarkus.agroal.runtime.DataSources] (main) Datasource <default> enables XA but transaction recovery is 
not enabled. Please enable transaction recovery by setting quarkus.transaction-manager.enable-recovery=true, otherwise 
data may be lost if the application is terminated abruptly
```
I tried setting the build setting `KC_TRANSACTION_XA_ENABLED=true` but it has no effect on the warning.

The message itself suggests another setting, `quarkus.transaction-manager.enable-recovery`. I don't know how to apply that, and I haven't seen it documented. How would I go about setting it?

I've seen a number unrelated issue with log output, that include the same warning. So it doesn't seem to be very rare. Is it important to fix?

**Comment by phoenix-ua** [comment]
> The message itself suggests another setting, quarkus.transaction-manager.enable-recovery. I don't know how to apply that, and I haven't seen it documented. How would I go about setting it?

We're upgrading from Keycloak 8.0.1 to 22.0.5 with PostgreSQL 12 and also encounter warnings when Keycloak starts, your warning is also there. The doc [Format for raw Quarkus properties](https://www.keycloak.org/server/configuration#_format_for_raw_quarkus_properties) helped me to make the warning goneю

But the funny thing that I still do not fully understand in which cases this transaction recovery mechanism can save us, given we're using a single Keycloak instance, and how to test it 😄

upd: This SO [answer](https://stackoverflow.com/questions/21109362/how-to-process-distributed-transaction-within-postgresql) enlightened me a bit about the XA transaction matter. Most likely it will be useful for those who run Keycloak in a multi-node clustered environment with some rdbms.
